File: musren-nexus-backend/backend/src/main/java/ke/co/musren/backend/service/AfricasTalkingService.java
package ke.co.musren.backend.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;

/**
 * Africa's Talking API integration.
 * Used for:
 *  - Airtime disbursement (affiliate withdrawal method: airtime)
 *  - Data bundle disbursement (affiliate withdrawal method: data)
 *  - Bulk SMS dispatch
 *
 * Sandbox: use username="sandbox" with any apiKey
 * Docs: https://developers.africastalking.com
 */
@Service
@Slf4j
public class AfricasTalkingService {

    private static final String BASE_URL = "https://api.africastalking.com";
    private static final String SANDBOX_URL = "https://api.sandbox.africastalking.com";

    @Value("${africastalking.username}")
    private String username;

    @Value("${africastalking.api-key}")
    private String apiKey;

    @Value("${africastalking.sender-id}")
    private String senderId;

    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public AfricasTalkingService(ObjectMapper objectMapper) {
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(10))
                .build();
        this.objectMapper = objectMapper;
    }

    private String apiBase() {
        return "sandbox".equals(username) ? SANDBOX_URL : BASE_URL;
    }

    /**
     * Sends airtime to a single phone number.
     *
     * @param phoneNumber  International format: +254712345678
     * @param amountKes    Amount in KES (e.g. 50 for KES 50)
     * @return Africa's Talking transaction ID
     */
    public String sendAirtime(String phoneNumber, double amountKes) throws IOException, InterruptedException {
        // AT expects amount as "KES X" string
        String recipientsJson = """
                [{"phoneNumber":"%s","amount":"KES %s"}]
                """.formatted(phoneNumber, amountKes).strip();

        String body = "username=" + encode(username)
                + "&recipients=" + encode(recipientsJson);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiBase() + "/version1/airtime/send"))
                .header("apiKey", apiKey)
                .header("Accept", "application/json")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .timeout(Duration.ofSeconds(15))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 201) {
            throw new IOException("AT airtime failed [%d]: %s".formatted(response.statusCode(), response.body()));
        }

        JsonNode node = objectMapper.readTree(response.body());
        JsonNode responses = node.path("responses");
        if (responses.isArray() && !responses.isEmpty()) {
            JsonNode first = responses.get(0);
            String status = first.path("status").asText();
            String transactionId = first.path("requestId").asText();
            log.info("AT airtime sent: status={}, txId={}, phone={}, amount=KES {}", status, transactionId, phoneNumber, amountKes);
            return transactionId;
        }

        throw new IOException("AT airtime: unexpected response: " + response.body());
    }

    /**
     * Sends an SMS to one or more recipients.
     *
     * @param message      Text body (max 160 chars per segment)
     * @param recipients   Comma-separated phone numbers in international format
     * @return AT messageId
     */
    public String sendSms(String message, String recipients) throws IOException, InterruptedException {
        String body = "username=" + encode(username)
                + "&to=" + encode(recipients)
                + "&message=" + encode(message)
                + "&from=" + encode(senderId);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiBase() + "/version1/messaging"))
                .header("apiKey", apiKey)
                .header("Accept", "application/json")
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(body))
                .timeout(Duration.ofSeconds(15))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 201) {
            throw new IOException("AT SMS failed [%d]: %s".formatted(response.statusCode(), response.body()));
        }

        JsonNode node = objectMapper.readTree(response.body());
        String messageId = node.path("SMSMessageData").path("Recipients").path(0).path("messageId").asText("-");
        log.info("AT SMS sent: messageId={}, to={}", messageId, recipients);
        return messageId;
    }

    /**
     * Sends a data bundle to a phone number.
     * NOTE: Africa's Talking data bundle API is carrier/country specific.
     * This sends airtime equivalent that the user can convert to data.
     * For direct data bundles, integrate with the specific carrier's API.
     *
     * @param phoneNumber  International format: +254712345678
     * @param amountKes    KES value of data to send
     */
    public String sendData(String phoneNumber, double amountKes) throws IOException, InterruptedException {
        // For Kenya, AT doesn't have a direct data bundle API — use airtime as proxy
        // The user can purchase data bundles with received airtime
        log.info("Data disbursement via airtime proxy: phone={}, amount=KES {}", phoneNumber, amountKes);
        return sendAirtime(phoneNumber, amountKes);
    }

    private String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }
}

File: musren-nexus-backend/api/app/Services/AfricasTalkingService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\Http;
use Illuminate\Support\Facades\Log;

class AfricasTalkingService
{
    private string $username;
    private string $apiKey;
    private string $senderId;
    private string $baseUrl;

    public function __construct()
    {
        $this->username = config('services.africastalking.username', 'sandbox');
        $this->apiKey   = config('services.africastalking.api_key');
        $this->senderId = config('services.africastalking.sender_id', 'MUSREN');
        $this->baseUrl  = $this->username === 'sandbox'
            ? 'https://api.sandbox.africastalking.com/version1'
            : 'https://api.africastalking.com/version1';
    }

    /**
     * Send SMS to one or more recipients.
     * @param string|array $to  Phone number(s) e.g. "+254700000001"
     */
    public function sendSms(string|array $to, string $message): array
    {
        $recipients = is_array($to) ? implode(',', $to) : $to;

        try {
            $response = Http::withHeaders([
                'apiKey' => $this->apiKey,
                'Accept' => 'application/json',
            ])->asForm()->post("{$this->baseUrl}/messaging", [
                'username'  => $this->username,
                'to'        => $recipients,
                'message'   => $message,
                'from'      => $this->senderId,
            ]);

            $data = $response->json();
            Log::info('AT SMS sent', ['recipients' => $recipients, 'response' => $data]);
            return ['success' => true, 'data' => $data];
        } catch (\Throwable $e) {
            Log::error('AT SMS failed', ['error' => $e->getMessage()]);
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }

    /**
     * Send airtime to a recipient.
     */
    public function sendAirtime(string $phone, string $amount, string $currency = 'KES'): array
    {
        try {
            $response = Http::withHeaders([
                'apiKey' => $this->apiKey,
                'Accept' => 'application/json',
            ])->asForm()->post("{$this->baseUrl}/airtime/send", [
                'username'    => $this->username,
                'recipients'  => json_encode([[
                    'phoneNumber' => $phone,
                    'amount'      => "{$currency} {$amount}",
                ]]),
            ]);

            $data = $response->json();
            Log::info('AT Airtime sent', ['phone' => $phone, 'amount' => $amount]);
            return ['success' => true, 'data' => $data];
        } catch (\Throwable $e) {
            Log::error('AT Airtime failed', ['error' => $e->getMessage()]);
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }
}
